applications/serial_lte_modem/src/slm_at_sms.c

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <logging/log.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <string.h>
+#include <modem/sms.h>
+#include "slm_util.h"
+
+LOG_MODULE_REGISTER(slm_sms, CONFIG_SLM_LOG_LEVEL);
+
+/**@brief SMS operations. */
+enum slm_sms_operation {
+	AT_SMS_STOP,
+	AT_SMS_START,
+	AT_SMS_SEND
+};
+
+static int sms_handle;
+
+/* global functions defined in different files */
+void rsp_send(const uint8_t *str, size_t len);
+
+/* global variable defined in different files */
+extern struct at_param_list at_param_list;
+extern char rsp_buf[CONFIG_SLM_SOCKET_RX_MAX * 2];
+
+static void sms_callback(struct sms_data *const data, void *context)
+{
+	static uint16_t ref_number;
+	static uint8_t total_msgs;
+	static uint8_t seq_number;
+
+	ARG_UNUSED(context);
+
+	if (data == NULL) {
+		LOG_WRN("NULL data");
+		return;
+	}
+
+	if (data->type == SMS_TYPE_DELIVER) {
+		struct sms_deliver_header *header = &data->header.deliver;
+
+		if (!header->concatenated.present) {
+			sprintf(rsp_buf, "\r\n#XSMS:\"%02d-%02d-%02d %02d:%02d:%02d\",\"",
+				header->time.year, header->time.month, header->time.day,
+				header->time.hour, header->time.minute, header->time.second);
+			strcat(rsp_buf, header->originating_address.address_str);
+			strcat(rsp_buf, "\",\"");
+			strcat(rsp_buf, data->payload);
+			strcat(rsp_buf, "\"\r\n");
+			rsp_send(rsp_buf, strlen(rsp_buf));
+		} else {
+			LOG_DBG("concatenated message %d, %d, %d",
+				header->concatenated.ref_number,
+				total_msgs = header->concatenated.total_msgs,
+				header->concatenated.seq_number);
+			/* ref_number and total_msgs should remain unchanged */
+			if (ref_number == 0) {
+				ref_number = header->concatenated.ref_number;
+			}
+			if (ref_number != header->concatenated.ref_number) {
+				LOG_ERR("SMS concatenated message ref_number error: %d, %d",
+					ref_number, header->concatenated.ref_number);
+				goto done;
+			}
+			if (total_msgs == 0) {
+				total_msgs = header->concatenated.total_msgs;
+			}
+			if (total_msgs != header->concatenated.total_msgs) {
+				LOG_ERR("SMS concatenated message total_msgs error: %d, %d",
+					total_msgs, header->concatenated.total_msgs);
+				goto done;
+			}
+			/* seq_number should start with 1 and increase by 1 each time */
```

Comment:
If you mean that concatenated SMS parts will arrive in the order of the sequence number, that's not true. They can arrive in random order and I've seen that happening in practice, actually surprisingly often.
Looks like code below won't work with random order message parts.

Reply:
I tested 3-trunk concatenated message and they are arriving in sequential order so I misunderstood the spec. I'll fix that based on your description. Thanks a lot to point it out!